### PR TITLE
Perception screen added for the SA

### PR DESCRIPTION
**Pull request by MohmadAyman**
### Description

This perception screen works as the following:
if (package has a 3d sensors config file) loads it into the screen
else load the default values from `moveit_setup_assistanat/resources/config/sensors_rgbd.yaml` -name is not final-.
when the user leaves the screen, saves the new config values to the `sensors_rgbd.yaml` to the package, if the user selected `none` from the plugin type drop down then no config file is saved.

### Checklist
- [x] **Required**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extended the tutorials / documentation, if necessary [reference](http://moveit.ros.org/documentation/contributing/)
- [x] Include a screenshot if changing a GUI
- [x] Optional: Created tests, which fail without this PR [reference](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/tests.html)
- [x] Optional: Decide if this should be cherry-picked to other current ROS branches (Indigo, Jade, Kinetic)

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Comment by MohmadAyman** [comment]
Is it Okay add 2 commits to the same PR? the second one have changes unrelated to the first one.

File: moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h [hunk 1]
```diff
@@ -370,6 +381,24 @@ class MoveItConfigData
    */
   std::vector<ROSControlConfig> getROSControllers();
 
+  /**
+   * \brief Used for adding a sensor plugin configuation prameter to the sensor plugin configuration parameter list
+   *
+   * @param parameter name
+   * @param parameter value
+   * @param parameter comment
```
Comment (davetcoleman) [inline moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h]: remove 'parameter' from each line

also, either add a real description for each parameter or remove the @param's part as it doesn't serve a purpose currently

File: moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h [hunk 2]
```diff
@@ -370,6 +381,24 @@ class MoveItConfigData
...
+   */
+  void addParameterToSensorPluginConfig(const std::string& name, const std::string& value = "",
+                                        const std::string& comment = "");
+
+  /** \brief This Function for clearing the sensor plugin configuration parameter list */
```
Comment (davetcoleman) [inline moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h]: More concise:
"This Function for clearing the..." -> "Clear the..."

File: moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h [hunk 3]
```diff
@@ -370,6 +381,24 @@ class MoveItConfigData
...
+  void clearSensorPluginConfig();
+
+  /**
+   * \brief Used for adding a sensor plugin configuation prameter to the sensor plugin configuration parameter list
```
Comment (davetcoleman) [inline moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h]: spelling "parameter"

File: moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h [hunk 4]
```diff
@@ -389,9 +418,11 @@ class MoveItConfigData
   // Private Vars
   // ******************************************************************************************
 
+  /// Sensor plugin configuration parameter list
+  std::vector<std::map<std::string, Parameter> > sensors_plugin_config_prameter_list_;
```
Comment (davetcoleman) [inline moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h]: spelling

File: moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h [hunk 5]
```diff
@@ -389,9 +418,11 @@ class MoveItConfigData
   // Private Vars
   // ******************************************************************************************
 
+  /// Sensor plugin configuration parameter list
```
Comment (davetcoleman) [inline moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h]: document what the vector does vs the map. this is a confusing type...

File: moveit_setup_assistant/resources/default_config/sensors_3d.yaml [hunk 6]
```diff
@@ -0,0 +1,20 @@
+# This YAML configuration file holds the default values used for configuring 3D sensors
```
Comment (davetcoleman) [inline moveit_setup_assistant/resources/default_config/sensors_3d.yaml]: Add comment:

     TODO: values may not be ideal defaults, original source unclear

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 7]
```diff
@@ -807,6 +804,49 @@ bool MoveItConfigData::outputROSControllersYAML(const std::string& file_path)
...
+  YAML::Emitter emitter;
+  emitter << YAML::BeginMap;
+
+  emitter << YAML::Key << "sensors";
+  emitter << YAML::Value << YAML::BeginSeq;
+
+  // Can we have more than one plugin config?
+  emitter << YAML::BeginMap;
+
+  // Make sure a sensor plugin was chosen from the perception screen
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: i don't understand what this does

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 8]
```diff
@@ -1360,6 +1400,101 @@ bool MoveItConfigData::inputSetupAssistantYAML(const std::string& file_path)
...
 
+// ******************************************************************************************
+// Input sensors_3d yaml file
+// ******************************************************************************************
+bool MoveItConfigData::input3DSensorsYAML(const std::string& default_file_path, const std::string& file_path)
+{
+  // Load file
+  std::ifstream default_input_stream(default_file_path.c_str());
+  if (!default_input_stream.good())
+  {
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: add ROS_ERROR?

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 9]
```diff
@@ -1360,6 +1400,101 @@ bool MoveItConfigData::inputSetupAssistantYAML(const std::string& file_path)
...
+    return false;
+  }
+
+  // Parse default paramertes values
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: "parameters"

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 10]
```diff
@@ -1360,6 +1400,101 @@ bool MoveItConfigData::inputSetupAssistantYAML(const std::string& file_path)
...
+    // Get sensors node
+    if (const YAML::Node& sensors_node = doc["sensors"])
+    {
+      if (sensors_node.IsSequence())
+      {
+        Parameter sensor_param;
+        std::map<std::string, Parameter> sensor_map;
+        for (std::size_t i = 0; i < sensors_node.size(); ++i)
+        {
+          if (const YAML::Node& sensor_node = sensors_node[i])
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: add a few more comments for all these nested ``if`` and ``for``

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 11]
```diff
@@ -1360,6 +1400,101 @@ bool MoveItConfigData::inputSetupAssistantYAML(const std::string& file_path)
...
+            }
+            sensors_plugin_config_prameter_list_.push_back(sensor_map);
+          }
+        }
+      }
+    }
+  }
+  catch (YAML::ParserException& e)  // Catch errors
+  {
+    ROS_ERROR_STREAM(e.what());
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: add default error: 

     ROS_ERROR_STREAM("Error parsing sensors yaml: " << e.what());

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 12]
```diff
@@ -1360,6 +1400,101 @@ bool MoveItConfigData::inputSetupAssistantYAML(const std::string& file_path)
...
+            sensors_plugin_config_prameter_list_.push_back(sensor_map);
+          }
+        }
+      }
+    }
+    return true;
+  }
+  catch (YAML::ParserException& e)  // Catch errors
+  {
+    ROS_ERROR_STREAM(e.what());
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: add error message also

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 13]
```diff
@@ -1439,4 +1574,34 @@ std::vector<ROSControlConfig> MoveItConfigData::getROSControllers()
   return ros_controllers_config_;
 }
 
-}  // namespace
\ No newline at end of file
+// Used to add a sensor plugin configuation prameter to the sensor plugin configuration parameter list
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: "parameter"

also add the row of stars above the comment :)

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 14]
```diff
@@ -1439,4 +1574,34 @@ std::vector<ROSControlConfig> MoveItConfigData::getROSControllers()
...
+void MoveItConfigData::addParameterToSensorPluginConfig(const std::string& name, const std::string& value,
+                                                        const std::string& comment)
+{
+  Parameter temp;
+  temp.name = name;
+  temp.value = value;
+  temp.comment = comment;
+
+  // Use index 0 since we only write one plugin
+  sensors_plugin_config_prameter_list_[0][name] = temp;
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp]: if you wanna learn some cool new C++11 stuff, try bracket initialization:

```
sensors_plugin_config_prameter_list_[0][name] = {
   .name = name,
   .value = value,
   .comment = comment };
```

@mcevoyandy am i just showing off?

File: moveit_setup_assistant/src/widgets/perception_widget.cpp [hunk 15]
```diff
@@ -0,0 +1,345 @@
...
+  : SetupScreenWidget(parent), config_data_(config_data)
+{
+  // Basic widget container
+  QVBoxLayout* layout = new QVBoxLayout();
+  layout->setAlignment(Qt::AlignTop);
+
+  // Top Header Area ------------------------------------------------
+
+  HeaderWidget* header =
+      new HeaderWidget("Perception Configuration", "Configure your 3D sensors to work with Moveit! ", this);
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/perception_widget.cpp]: title: "3D Perception Sensor Configuration"

File: moveit_setup_assistant/src/widgets/perception_widget.cpp [hunk 16]
```diff
@@ -0,0 +1,345 @@
...
+  horizontal_line->setStyleSheet(QString("background-color: #c0c0c0;"));
+  layout->addWidget(horizontal_line);
+
+  // Add spacing
+  QSpacerItem* blank_space = new QSpacerItem(1, 4);
+  layout->addSpacerItem(blank_space);
+
+  // Plugin type combo box
+  QLabel* plugin_field_title = new QLabel(this);
+  plugin_field_title->setText("Type of the sensor plugin to configure:");
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/perception_widget.cpp]: "Optionally choose a type of 3D sensor plugin to configure:"

File: moveit_setup_assistant/src/widgets/perception_widget.cpp [hunk 17]
```diff
@@ -0,0 +1,345 @@
...
+  // Filtered Cloud Topic
+  depth_max_update_rate_field_ = new QLineEdit(this);
+  depth_max_update_rate_field_->setMaximumWidth(400);
+  depth_map_form_layout->addRow("Max Update Rate:", depth_max_update_rate_field_);
+
+  depth_map_group_->setLayout(depth_map_form_layout);
+  layout->addWidget(depth_map_group_);
+
+  // Perception documentation page
+  QLabel* perception_documentation_hyperlink = new QLabel(this);
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/perception_widget.cpp]: move to top as we discussed please. or move to bottom but don't have it jump around :)

File: moveit_setup_assistant/src/widgets/perception_widget.cpp [hunk 18]
```diff
@@ -0,0 +1,345 @@
...
+    }
+  }
+
+  // If no sensor config exists, default to None
+  if (sensors_vec_map.size() == 2)
+  {
+    sensor_plugin_field_->setCurrentIndex(0);
+  }
+}
+
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/perception_widget.cpp]: great work with this file!!

File: moveit_setup_assistant/src/widgets/setup_assistant_widget.cpp [hunk 19]
```diff
@@ -90,22 +90,22 @@ SetupAssistantWidget::SetupAssistantWidget(QWidget* parent, boost::program_optio
   // Screens --------------------------------------------------------
 
   // Start Screen
-  ssw_ = new StartScreenWidget(this, config_data_);
-  ssw_->setSizePolicy(QSizePolicy::Preferred, QSizePolicy::Preferred);
-  connect(ssw_, SIGNAL(readyToProgress()), this, SLOT(progressPastStartScreen()));
-  connect(ssw_, SIGNAL(loadRviz()), this, SLOT(loadRviz()));
-  main_content_->addWidget(ssw_);
+  start_screen_widget_ = new StartScreenWidget(this, config_data_);
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/setup_assistant_widget.cpp]: this should really be in a separate PR...

File: moveit_setup_assistant/src/widgets/setup_assistant_widget.h [hunk 20]
```diff
@@ -213,15 +214,16 @@ private Q_SLOTS:
...
+  StartScreenWidget* start_screen_widget_;
+  DefaultCollisionsWidget* default_collisions_widget_;
+  PlanningGroupsWidget* planning_groups_widget_;
+  RobotPosesWidget* robot_poses_widget_;
+  EndEffectorsWidget* end_effectors_widget_;
+  VirtualJointsWidget* virtual_joints_widget_;
+  PassiveJointsWidget* passive_joints_widget_;
+  PerceptionWidget* perception_widget_;
+  AuthorInformationWidget* author_information_widget_;
+  ConfigurationFilesWidget* configuration_files_widget_;
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/setup_assistant_widget.h]: thanks for doing this!

File: moveit_setup_assistant/src/widgets/start_screen_widget.cpp [hunk 21]
```diff
@@ -421,6 +421,23 @@ bool StartScreenWidget::loadExistingFiles()
   fs::path kinematics_yaml_path = config_data_->config_pkg_path_;
   kinematics_yaml_path /= "config/kinematics.yaml";
 
+  // Load sensors_3d yaml file if available --------------------------------------------------
+  fs::path sensors_3d_yaml_path = config_data_->config_pkg_path_;
+  sensors_3d_yaml_path /= "config/sensors_3d.yaml";
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/start_screen_widget.cpp]: add note inside the sensors_3d.yaml files that the filename should not be changed or else the MSA will not detect it

File: moveit_setup_assistant/test/moveit_config_data_test.cpp [hunk 22]
```diff
@@ -119,6 +119,32 @@ TEST_F(MoveItConfigData, ReadingControllers)
...
 }
 
+// This tests parsing of sensors_rgbd.yaml
+TEST_F(MoveItConfigData, ReadingSensorsConfig)
+{
+  // Contains all the config data for the setup assistant
+  moveit_setup_assistant::MoveItConfigDataPtr config_data_;
+  config_data_.reset(new moveit_setup_assistant::MoveItConfigData());
+
+  boost::filesystem::path setup_assistnat_path(config_data_->setup_assistant_path_);
```
Comment (davetcoleman) [inline moveit_setup_assistant/test/moveit_config_data_test.cpp]: *assistant (spelling)

File: moveit_setup_assistant/CMakeLists.txt [hunk 23]
```diff
@@ -146,9 +148,8 @@ install(DIRECTORY templates DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION})
 if(CATKIN_ENABLE_TESTING)
   find_package(moveit_resources REQUIRED)
   include_directories(${moveit_resources_INCLUDE_DIRS})
-
-  catkin_add_gtest(test_writing_ros_controllers test/moveit_config_data_test.cpp)
-  target_link_libraries(test_writing_ros_controllers  ${PROJECT_NAME}_tools
+  catkin_add_gtest(test_reading_writing_config test/moveit_config_data_test.cpp)
+  target_link_libraries(test_reading_writing_config  ${PROJECT_NAME}_tools
```
[the remark this answers is not in the document]
Comment (MohmadAyman) [reply moveit_setup_assistant/CMakeLists.txt]: This PR adds a test to the `moveit_config_data_test.cpp`, so I renamed the test to be sth more generic.

File: moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h [hunk 24]
```diff
@@ -79,9 +79,9 @@ struct GroupMetaData
 };
 
 /**
- * Planning parameters which may be set in the config files
+ * Parameters which may be set in the config files
```
[the remark this answers is not in the document]
Comment (MohmadAyman) [reply moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h]: "Reusable parameter struct which may be used in reading and writing configuration files"

as that is the main usage of this struct, is that okay?

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp (hunk 14 above)
Comment (MohmadAyman) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp]: Oh Thats Nice!

**Comment by MohmadAyman** [comment]
@davetcoleman Taken care of all the comments, feel free to review :)

File: moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h [hunk 25]
```diff
@@ -79,9 +79,9 @@ struct GroupMetaData
 };
 
 /**
- * Planning parameters which may be set in the config files
+ * Reusable parameter struct which may be used in reading and writing configuration files
  */
-struct OmplPlanningParameter
+struct GenericParameter
```
Comment (davetcoleman) [inline moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h]: this should actually be a class since it has member functions (see GStyle guide)

File: moveit_setup_assistant/resources/default_config/sensors_3d.yaml [hunk 26]
```diff
@@ -0,0 +1,22 @@
+# This YAML configuration file holds the default values used for configuring 3D sensors.
+# The name of this file shouldn't be changed, or else the Setup Assistant won't detect it.
```
[the remark this answers is not in the document]
Comment (Ridhwanluthra) [reply moveit_setup_assistant/resources/default_config/sensors_3d.yaml]: The name of the yaml file "sensors_3d.yaml"

File: moveit_setup_assistant/resources/default_config/sensors_3d.yaml [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+sensors:
+  - sensor_plugin: occupancy_map_monitor/PointCloudOctomapUpdater
+    point_cloud_topic: /head_mount_kinect/depth_registered/points
+    max_range: 5.0
+    point_subsample: 1
+    padding_offset: 0.1
+    padding_scale: 1.0
+    max_update_rate: 1.0
+    filtered_cloud_topic: filtered_cloud
+  - sensor_plugin: occupancy_map_monitor/DepthImageOctomapUpdater
```
Comment (Ridhwanluthra) [inline moveit_setup_assistant/resources/default_config/sensors_3d.yaml]: Won't have both of them in the same file end up loading both of them?

Comment (MohmadAyman) [reply moveit_setup_assistant/resources/default_config/sensors_3d.yaml]: Yes, but we don't put this file on the parameter server, we only use it for holding the default values.

File: moveit_setup_assistant/resources/default_config/sensors_3d.yaml (hunk 26 above)
[the remark this answers is not in the document]
Comment (MohmadAyman) [reply moveit_setup_assistant/resources/default_config/sensors_3d.yaml]: I'm not sure if I understand correctly but the name of the yaml file "sensors_3d.yaml" doesn't change if thats what you're asking.

File: moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h (hunk 25 above)
Comment (MohmadAyman) [reply moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h]: Yes, i'll be working on that :+1:

Comment (MohmadAyman) [reply moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h]: I think this struct is just passive.

I mean the member functions that uses this struct are `addGenericParameterToSensorPluginConfig()` and `SensorPluginConfig` is a member of the `moveit_config_data`, so if we made a class all it functions will be just `get` and `set` functions and the `addGenericParamereter` function would still be a function in the `movit_config_data` class.
Same with `getSensorPluginConfig()`.
Did I get it correctly?

Comment (MohmadAyman) [reply moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h]: The controllers screen is also ready, but if this one really needs the struct to be class i'll modify it before I open a PR for it.

Comment (MohmadAyman) [reply moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h]: I've made two commits, the first one uses a struct and in the second one I've converted it into a class.

**Comment by davetcoleman** [comment]
I need a second +1 from someone @mcevoyandy

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/sensor_manager.launch.xml [hunk 28]
```diff
@@ -2,6 +2,9 @@
 
   <!-- This file makes it easy to include the settings for sensor managers -->
 
+  <!-- Params for 3D sensors config -->
+  <rosparam command="load" file="$(find [GENERATED_PACKAGE_NAME])/config/sensors_3d.yaml" />
```
Comment (gavanderhoorn) [inline moveit_setup_assistant/templates/moveit_config_pkg_template/launch/sensor_manager.launch.xml]: Late, but what was the rationale for not using the "robot specific sensor manager" `.launch.xml` infrastructure used further on in this file to load the `.yaml`?

This seems to have deviated from how this worked previously, and seems to make `[ROBOT_NAME]_moveit_sensor_manager.launch.xml` redundant (but that file is still generated by the Setup Assistant).
